File: code_java/src/main/java/jp/co/pm/ai/desktop/debug/AgentDebugLog.java
package jp.co.pm.ai.desktop.debug;

import java.lang.management.ManagementFactory;
import java.lang.management.MemoryUsage;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Comparator;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.fasterxml.jackson.databind.ObjectMapper;

import jp.co.pm.ai.desktop.config.AppPaths;

/**
 * NDJSON append for Cursor debug sessions. Resolves log path across Windows / WSL / fat JAR layouts.
 *
 * <p>Project convention: {@code .cursor/rules/agent-debug-ndjson-logging.mdc}（モノレポでは親
 * {@code 工程管理AIプロジェクト_JAVA/.cursor/rules/agent-debug-ndjson-logging.mdc} が正本）。
 *
 * <p>Resolution order (first hit wins):
 *
 * <ol>
 *   <li>{@code CURSOR_DEBUG_LOG} or {@code PM_AI_DEBUG_LOG} (absolute path to log file)</li>
 *   <li>UI env {@link AppPaths#KEY_PM_AI_CURSOR_DEBUG_LOG}</li>
 *   <li>{@code parent(resolveRepoRoot) / .cursor / debug-&lt;sessionId&gt;.log} (workspace {@code .cursor} when repo is
 *       {@code Production-Control-System})</li>
 *   <li>{@code resolveRepoRoot / .cursor / debug-&lt;sessionId&gt;.log}</li>
 * </ol>
 *
 * <p>If the primary path cannot be written, retries under {@code user.home/.cursor/}.
 *
 * <p>WSL コンソールから {@code powershell.exe} で Windows JVM を起動した場合、NDJSON はドライブレター経由（{@code C:\...}）で
 * 書かれることがある。Cursor がワークスペースを {@code /mnt/c/...}（WSL）として開いていると、エージェントがログを見つけにくい。
 * そのため一次書き込み成功後にミラー追記する。詳細は {@code .cursor/rules/agent-debug-wsl-windows-mirror.mdc}。
 */
public final class AgentDebugLog {

    public static final String DEFAULT_SESSION_ID = "e04a1d";

    private static final ObjectMapper JSON = new ObjectMapper();

    /** Lazily resolved when {@code WSL_DISTRO_NAME} / {@code PM_AI_WSL_DISTRO} are unset (powershell.exe 子プロセスではよく未設定）。 */
    private static volatile boolean wslDistroDiscoveryAttempted;

    private static volatile String cachedDiscoveredWslDistro;

    private AgentDebugLog() {}

    /** Heap used/max in MiB for debug NDJSON payloads (session tooling). */
    public static Map<String, Object> debugHeapMap() {
        MemoryUsage h = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("heapUsedMiB", h.getUsed() / (1024L * 1024L));
        long max = h.getMax();
        m.put("heapMaxMiB", max > 0 ? max / (1024L * 1024L) : -1L);
        return m;
    }

    public static Path resolveNdjsonPath(Map<String, String> ui, String sessionId) {
        String id =
                sessionId == null || sessionId.isBlank()
                        ? DEFAULT_SESSION_ID
                        : sessionId.trim();
        String fileName = "debug-" + id + ".log";

        Path envPath = pathFromEnvOverride();
        if (envPath != null) {
            return envPath;
        }

        Map<String, String> u = ui != null ? ui : Map.of();
        String uiDbg = trim(u.get("PM_AI_DEBUG_LOG"));
        if (!uiDbg.isEmpty()) {
            return Path.of(uiDbg).toAbsolutePath().normalize();
        }
        String uiPath = trim(u.get(AppPaths.KEY_PM_AI_CURSOR_DEBUG_LOG));
        if (!uiPath.isEmpty()) {
            return Path.of(uiPath).toAbsolutePath().normalize();
        }

        Path repo = AppPaths.resolveRepoRoot(u);
        /*
         * Nested clone: repo leaf is Production-Control-System → workspace .cursor is parent(repo)/.cursor
         * (see agent-debug-ndjson-logging.mdc). Flat repo: repo/.cursor (parent would be drive root — wrong).
         */
        Path parent = repo.getParent();
        if (parent != null && isProductionControlSystemRepoLeaf(repo)) {
            return parent.resolve(".cursor").resolve(fileName).toAbsolutePath().normalize();
        }
        return repo.resolve(".cursor").resolve(fileName).toAbsolutePath().normalize();
    }

    private static boolean isProductionControlSystemRepoLeaf(Path repo) {
        if (repo == null) {
            return false;
        }
        Path leaf = repo.getFileName();
        return leaf != null
                && "Production-Control-System".equalsIgnoreCase(leaf.toString());
    }

    /**
     * Appends one UTF-8 line; creates parent directories. Falls back to {@code user.home/.cursor/} only if primary
     * fails.
     *
     * @return file path actually written, or {@code null} if both attempts failed
     */
    public static Path appendNdjsonLine(Map<String, String> ui, String sessionId, String jsonLine) {
        String line = jsonLine.endsWith("\n") ? jsonLine : jsonLine + "\n";
        Path primary = resolveNdjsonPath(ui, sessionId);
        if (writeUtf8Append(primary, line)) {
            appendMirrors(primary, line, ui);
            return primary;
        }
        String id =
                sessionId == null || sessionId.isBlank()
                        ? DEFAULT_SESSION_ID
                        : sessionId.trim();
        Path fallback =
                Path.of(System.getProperty("user.home", "."))
                        .resolve(".cursor")
                        .resolve("debug-" + id + ".log")
                        .toAbsolutePath()
                        .normalize();
        if (writeUtf8Append(fallback, line)) {
            appendMirrors(fallback, line, ui);
            return fallback;
        }
        return null;
    }

    /** Package-private: mirror path string for tests ({@code C:\...} → {@code \\wsl$\distro\mnt\c\...}). */
    static String buildWslUncPathString(String windowsNormalizedAbsolute, String distro) {
        return buildWslUncPathString(windowsNormalizedAbsolute, distro, "\\\\wsl$\\");
    }

    /**
     * @param uncRootPrefix UNC のルート（{@code \\\\wsl$\\} または {@code \\\\wsl.localhost\\}）。末尾はバックスラッシュを付ける。
     */
    static String buildWslUncPathString(
            String windowsNormalizedAbsolute, String distro, String uncRootPrefix) {
        if (distro == null || distro.isBlank()) {
            return null;
        }
        if (windowsNormalizedAbsolute == null || windowsNormalizedAbsolute.length() < 3) {
            return null;
        }
        char dl = windowsNormalizedAbsolute.charAt(0);
        if (!Character.isLetter(dl) || windowsNormalizedAbsolute.charAt(1) != ':') {
            return null;
        }
        String tail = windowsNormalizedAbsolute.substring(2).replace('/', '\\');
        if (!tail.startsWith("\\")) {
            tail = "\\" + tail;
        }
        String root = uncRootPrefix != null ? uncRootPrefix : "\\\\wsl$\\";
        return root
                + distro.trim()
                + "\\mnt\\"
                + Character.toLowerCase(dl)
                + tail;
    }

    private static boolean isWindowsOs() {
        return System.getProperty("os.name", "").toLowerCase(Locale.ROOT).contains("windows");
    }

    private static boolean wslUncMirrorEnabled() {
        String v = trim(System.getenv("PM_AI_DEBUG_LOG_WSL_UNC"));
        if (v.isEmpty()) {
            return true;
        }
        return !(v.equals("0") || v.equalsIgnoreCase("false") || v.equalsIgnoreCase("off"));
    }

    private static String resolveWslDistroName() {
        String d = trim(System.getenv("PM_AI_WSL_DISTRO"));
        if (!d.isEmpty()) {
            return d;
        }
        d = trim(System.getenv("WSL_DISTRO_NAME"));
        if (!d.isEmpty()) {
            return d;
        }
        return discoverWslDistroNameCached();
    }

    private static String discoverWslDistroNameCached() {
        if (wslDistroDiscoveryAttempted) {
            return cachedDiscoveredWslDistro;
        }
        synchronized (AgentDebugLog.class) {
            if (wslDistroDiscoveryAttempted) {
                return cachedDiscoveredWslDistro;
            }
            wslDistroDiscoveryAttempted = true;
            cachedDiscoveredWslDistro = discoverWslDistroByListingShare();
            return cachedDiscoveredWslDistro;
        }
    }

    /**
     * {@code \\\\wsl$\\} を列挙し、docker-desktop 等を除いて優先する（名前に {@code ubuntu} を含むものを優先）。
     */
    private static String discoverWslDistroByListingShare() {
        if (!isWindowsOs()) {
            return null;
        }
        try (Stream<Path> stream = Files.list(Path.of("\\\\wsl$\\"))) {
            List<String> names =
                    stream.filter(
                                    p -> {
                                        try {
                                            return Files.isDirectory(p);
                                        } catch (Throwable ignored) {
                                            return false;
                                        }
                                    })
                            .map(p -> p.getFileName().toString())
                            .filter(AgentDebugLog::isPlausibleUserWslDistroFolder)
                            .sorted()
                            .collect(Collectors.toList());
            Optional<String> ubuntu =
                    names.stream()
                            .filter(n -> n.toLowerCase(Locale.ROOT).contains("ubuntu"))
                            .min(Comparator.naturalOrder());
            return ubuntu.orElseGet(() -> names.isEmpty() ? null : names.get(0));
        } catch (Throwable ignored) {
            return null;
        }
    }

    private static boolean isPlausibleUserWslDistroFolder(String name) {
        if (name == null || name.isBlank()) {
            return false;
        }
        String lower = name.toLowerCase(Locale.ROOT);
        if (lower.equals("docker-desktop")
                || lower.startsWith("rancher-desktop")
                || lower.contains("podman")) {
            return false;
        }
        return true;
    }

    private static Path windowsDrivePathToWslUnc(Path absoluteDrivePath, String distro, String uncRoot) {
        String unc = buildWslUncPathString(absoluteDrivePath.normalize().toString(), distro, uncRoot);
        return unc != null ? Path.of(unc) : null;
    }

    private static void appendMirrors(Path writtenPath, String line, Map<String, String> ui) {
        for (Path mirror : resolveMirrorTargets(writtenPath, ui)) {
            if (mirror == null) {
                continue;
            }
            Path n = mirror.toAbsolutePath().normalize();
            Path w = writtenPath.toAbsolutePath().normalize();
            if (n.equals(w)) {
                continue;
            }
            try {
                if (Files.exists(w) && Files.exists(n) && Files.isSameFile(w, n)) {
                    continue;
                }
            } catch (Throwable ignored) {
                // attempt mirror write
            }
            writeUtf8Append(n, line);
        }
    }

    private static List<Path> resolveMirrorTargets(Path primaryWritten, Map<String, String> ui) {
        Set<Path> out = new LinkedHashSet<>();
        Path explicit = mirrorPathFromEnv(ui);
        if (explicit != null) {
            out.add(explicit);
        }
        if (isWindowsOs() && wslUncMirrorEnabled()) {
            String distro = resolveWslDistroName();
            if (distro != null) {
                Path uncWsl = windowsDrivePathToWslUnc(primaryWritten, distro, "\\\\wsl$\\");
                if (uncWsl != null) {
                    out.add(uncWsl);
                }
                Path uncLh = windowsDrivePathToWslUnc(primaryWritten, distro, "\\\\wsl.localhost\\");
                if (uncLh != null) {
                    out.add(uncLh);
                }
            }
        }
        return new ArrayList<>(out);
    }

    private static Path mirrorPathFromEnv(Map<String, String> ui) {
        String v = trim(System.getenv("PM_AI_DEBUG_LOG_MIRROR"));
        if (v.isEmpty() && ui != null) {
            v = trim(ui.get(AppPaths.KEY_PM_AI_DEBUG_LOG_MIRROR));
        }
        return v.isEmpty() ? null : Path.of(v).toAbsolutePath().normalize();
    }

    private static boolean writeUtf8Append(Path path, String line) {
        try {
            Path dir = path.getParent();
            if (dir != null) {
                Files.createDirectories(dir);
            }
            Files.writeString(
                    path,
                    line,
                    StandardCharsets.UTF_8,
                    StandardOpenOption.CREATE,
                    StandardOpenOption.APPEND);
            return true;
        } catch (Throwable ignored) {
            return false;
        }
    }

    private static Path pathFromEnvOverride() {
        String v = trim(System.getenv("CURSOR_DEBUG_LOG"));
        if (!v.isEmpty()) {
            return Path.of(v).toAbsolutePath().normalize();
        }
        v = trim(System.getenv("PM_AI_DEBUG_LOG"));
        if (!v.isEmpty()) {
            return Path.of(v).toAbsolutePath().normalize();
        }
        return null;
    }

    private static String trim(String s) {
        return s != null ? s.trim() : "";
    }

    /**
     * One NDJSON line with session / hypothesis / data (debug sessions). Does not log secrets; keep {@code data} small.
     */
    public static void appendStructured(
            Map<String, String> ui,
            String sessionId,
            String hypothesisId,
            String location,
            String message,
            Map<String, ?> data) {
        try {
            Map<String, Object> line = new LinkedHashMap<>();
            String sid =
                    sessionId == null || sessionId.isBlank()
                            ? DEFAULT_SESSION_ID
                            : sessionId.trim();
            line.put("sessionId", sid);
            line.put("hypothesisId", hypothesisId);
            line.put("location", location);
            line.put("message", message);
            line.put("data", data != null ? data : Map.of());
            line.put("timestamp", System.currentTimeMillis());
            String json = JSON.writeValueAsString(line);
            appendNdjsonLine(ui, sid, json);
        } catch (Throwable ignored) {
            // debug-only
        }
    }

    /**
     * 配台試行など子プロセス NDJSON と JVM 側 {@link #appendStructured} の {@code sessionId} を揃える。
     *
     * <p>優先: UI の {@code PM_AI_AGENT_DEBUG_SESSION} → OS の同キー → {@code CURSOR_DEBUG_SESSION_ID} →
     * 既定（現在の Cursor デバッグ会話 ID。会話ごとに変える場合は環境変数タブで上書き）。
     */
    public static String resolveDispatchTrialSessionId(Map<String, String> ui) {
        if (ui != null) {
            String s = trim(ui.get("PM_AI_AGENT_DEBUG_SESSION"));
            if (!s.isEmpty()) {
                return s;
            }
        }
        String s2 = trim(System.getenv("PM_AI_AGENT_DEBUG_SESSION"));
        if (!s2.isEmpty()) {
            return s2;
        }
        String s3 = trim(System.getenv("CURSOR_DEBUG_SESSION_ID"));
        if (!s3.isEmpty()) {
            return s3;
        }
        return DEFAULT_SESSION_ID;
    }

    /**
     * Python 子プロセス（段階2・配台試行）向けに {@code PM_AI_AGENT_DEBUG_SESSION} と
     * {@code PM_AI_DEBUG_LOG} を解決して {@code env} に書き込む。環境変数タブで既に非空なら上書きしない。
     *
     * <p>正本: {@code .cursor/rules/agent-debug-ndjson-logging.mdc}
     */
    public static void overlayPythonChildDebugEnv(Map<String, String> env) {
        if (env == null) {
            return;
        }
        String sid = resolveDispatchTrialSessionId(env);
        env.put("PM_AI_AGENT_DEBUG_SESSION", sid);
        String dbg = trim(env.get("PM_AI_DEBUG_LOG"));
        if (dbg.isEmpty()) {
            dbg = trim(env.get(AppPaths.KEY_PM_AI_CURSOR_DEBUG_LOG));
        }
        if (dbg.isEmpty()) {
            env.put(
                    "PM_AI_DEBUG_LOG",
                    resolveNdjsonPath(env, sid).toAbsolutePath().toString());
        }
    }
}

File: code/python/planning_core/agent_debug_log.py
# -*- coding: utf-8 -*-
"""
Agent デバッグ用 NDJSON 追記（Python 側）。

Java の ``jp.co.pm.ai.desktop.debug.AgentDebugLog`` と同一の解決順・フォールバック方針に揃える。
正本ルール: ``.cursor/rules/agent-debug-ndjson-logging.mdc`` / ``agent-debug-wsl-windows-mirror.mdc``。
"""

from __future__ import annotations

import json
import os
import pathlib
import re
import sys
import time
from typing import Any, Mapping

# Java AgentDebugLog.DEFAULT_SESSION_ID と一致
DEFAULT_AGENT_DEBUG_SESSION_ID = "e04a1d"

KEY_PM_AI_REPO_ROOT = "PM_AI_REPO_ROOT"
KEY_PM_AI_CURSOR_DEBUG_LOG = "PM_AI_CURSOR_DEBUG_LOG"
KEY_PM_AI_DEBUG_LOG_MIRROR = "PM_AI_DEBUG_LOG_MIRROR"

_PRODUCTION_CONTROL_SYSTEM = "Production-Control-System"


def _trim(s: str | None) -> str:
    return (s or "").strip()


def ui_env_from_os() -> dict[str, str]:
    """Java の ``collectUiEnv()`` 相当に近い最小マップ（ログパス解決用）。"""
    keys = (
        KEY_PM_AI_REPO_ROOT,
        KEY_PM_AI_CURSOR_DEBUG_LOG,
        KEY_PM_AI_DEBUG_LOG_MIRROR,
    )
    return {k: _trim(os.environ.get(k)) for k in keys if _trim(os.environ.get(k))}


def resolve_agent_debug_session_id() -> str:
    for k in ("PM_AI_AGENT_DEBUG_SESSION_ID", "CURSOR_DEBUG_SESSION_ID"):
        v = _trim(os.environ.get(k))
        if v:
            return v
    return DEFAULT_AGENT_DEBUG_SESSION_ID


def _resolve_repo_root(ui: Mapping[str, str] | None) -> pathlib.Path:
    u = dict(ui or {})
    r = _trim(u.get(KEY_PM_AI_REPO_ROOT))
    if r:
        return pathlib.Path(r).resolve()
    here = pathlib.Path(__file__).resolve()
    code = here.parents[2]
    repo = code.parent
    return repo if repo is not None else code


def _is_production_control_system_repo_leaf(repo: pathlib.Path) -> bool:
    try:
        return repo.name.lower() == _PRODUCTION_CONTROL_SYSTEM.lower()
    except Exception:
        return False


def resolve_ndjson_path(ui: Mapping[str, str] | None, session_id: str | None) -> pathlib.Path:
    """
    ``AgentDebugLog.resolveNdjsonPath`` と同一の first-hit 順。

    1. ``CURSOR_DEBUG_LOG`` / ``PM_AI_DEBUG_LOG``（ログファイルへの絶対パス）
    2. ``ui`` / 環境の ``PM_AI_CURSOR_DEBUG_LOG``
    3. ネストクローン: ``parent(repo)/.cursor/debug-<session>.log``
    4. フラット: ``repo/.cursor/debug-<session>.log``
    """
    sid = _trim(session_id) or DEFAULT_AGENT_DEBUG_SESSION_ID
    name = f"debug-{sid}.log"
    v = _trim(os.environ.get("CURSOR_DEBUG_LOG"))
    if not v:
        v = _trim(os.environ.get("PM_AI_DEBUG_LOG"))
    if v:
        return pathlib.Path(v).resolve()
    u = dict(ui or {})
    ui_path = _trim(u.get(KEY_PM_AI_CURSOR_DEBUG_LOG))
    if not ui_path:
        ui_path = _trim(os.environ.get(KEY_PM_AI_CURSOR_DEBUG_LOG))
    if ui_path:
        return pathlib.Path(ui_path).resolve()
    repo = _resolve_repo_root(u)
    if _is_production_control_system_repo_leaf(repo):
        par = repo.parent
        if par is not None:
            return (par / ".cursor" / name).resolve()
    return (repo / ".cursor" / name).resolve()


def _write_utf8_append(path: pathlib.Path, line: str) -> bool:
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        with open(path, "a", encoding="utf-8") as f:
            f.write(line)
        return True
    except OSError:
        return False


def _build_wsl_unc_path(windows_abs: str, distro: str, unc_root: str) -> str | None:
    if not distro or len(windows_abs) < 3:
        return None
    dl = windows_abs[0]
    if not (dl.isalpha() and windows_abs[1] == ":"):
        return None
    tail = windows_abs[2:].replace("/", "\\")
    if not tail.startswith("\\"):
        tail = "\\" + tail
    root = unc_root if unc_root.endswith("\\") else unc_root + "\\"
    return f"{root}{distro.strip()}\\mnt\\{dl.lower()}{tail}"


def _wsl_unc_mirror_enabled() -> bool:
    v = _trim(os.environ.get("PM_AI_DEBUG_LOG_WSL_UNC"))
    if not v:
        return True
    return v not in ("0", "false", "False", "off", "OFF")


def _resolve_wsl_distro_name() -> str | None:
    d = _trim(os.environ.get("PM_AI_WSL_DISTRO")) or _trim(os.environ.get("WSL_DISTRO_NAME"))
    if d:
        return d
    if sys.platform != "win32":
        return None
    try:
        root = pathlib.Path("//wsl$/")
        if not root.is_dir():
            return None
        names: list[str] = []
        for p in root.iterdir():
            if not p.is_dir():
                continue
            n = p.name
            low = n.lower()
            if low == "docker-desktop" or low.startswith("rancher-desktop") or "podman" in low:
                continue
            names.append(n)
        names.sort()
        for n in names:
            if "ubuntu" in n.lower():
                return n
        return names[0] if names else None
    except OSError:
        return None


def _mirror_targets(primary: pathlib.Path, ui: Mapping[str, str] | None) -> list[pathlib.Path]:
    out: list[pathlib.Path] = []
    u = dict(ui or {})
    m = _trim(os.environ.get(KEY_PM_AI_DEBUG_LOG_MIRROR)) or _trim(
        u.get(KEY_PM_AI_DEBUG_LOG_MIRROR)
    )
    if m:
        out.append(pathlib.Path(m).resolve())
    if sys.platform != "win32" or not _wsl_unc_mirror_enabled():
        return out
    distro = _resolve_wsl_distro_name()
    if not distro:
        return out
    s = str(primary.resolve())
    for unc_root in ("\\\\wsl$\\", "\\\\wsl.localhost\\"):
        unc = _build_wsl_unc_path(s, distro, unc_root)
        if unc:
            try:
                out.append(pathlib.Path(unc))
            except Exception:
                pass
    return out


def _append_mirrors(primary: pathlib.Path, line: str, ui: Mapping[str, str] | None) -> None:
    w = primary.resolve()
    for m in _mirror_targets(primary, ui):
        try:
            n = m.resolve()
        except Exception:
            continue
        if n == w:
            continue
        try:
            if w.exists() and n.exists() and w.samefile(n):
                continue
        except OSError:
            pass
        _write_utf8_append(n, line)


def append_ndjson_line(
    ui: Mapping[str, str] | None,
    session_id: str | None,
    json_line: str,
) -> pathlib.Path | None:
    """
    1 行 UTF-8 追記。Java ``AgentDebugLog.appendNdjsonLine`` 相当。
    主パス失敗時は ``user.home/.cursor/debug-<session>.log`` にフォールバック。
    """
    line = json_line if json_line.endswith("\n") else json_line + "\n"
    sid = _trim(session_id) or DEFAULT_AGENT_DEBUG_SESSION_ID
    primary = resolve_ndjson_path(ui, sid)
    if _write_utf8_append(primary, line):
        _append_mirrors(primary, line, ui)
        return primary
    home = pathlib.Path.home() / ".cursor" / f"debug-{sid}.log"
    if _write_utf8_append(home, line):
        _append_mirrors(home, line, ui)
        return home
    return None


def append_structured(
    ui: Mapping[str, str] | None,
    session_id: str | None,
    hypothesis_id: str,
    location: str,
    message: str,
    data: Mapping[str, Any] | None,
) -> pathlib.Path | None:
    """Java ``AgentDebugLog.appendStructured`` と同形の 1 行 NDJSON。"""
    sid = _trim(session_id) or DEFAULT_AGENT_DEBUG_SESSION_ID
    try:
        payload: dict[str, Any] = {
            "sessionId": sid,
            "hypothesisId": hypothesis_id,
            "location": location,
            "message": message,
            "data": dict(data) if data is not None else {},
            "timestamp": int(time.time() * 1000),
        }
        return append_ndjson_line(ui, sid, json.dumps(payload, ensure_ascii=False))
    except Exception:
        return None


def append_interactive_core_probe(
    hypothesis_id: str,
    location: str,
    message: str,
    data: Mapping[str, Any] | None = None,
) -> pathlib.Path | None:
    """
    ``planning_core._core`` インタラクティブ試行デバッグ用。
    ``PM_AI_AGENT_DEBUG_SESSION_ID`` / ``CURSOR_DEBUG_SESSION_ID`` で sessionId を上書き可。
    """
    return append_structured(
        ui_env_from_os(),
        resolve_agent_debug_session_id(),
        hypothesis_id,
        location,
        message,
        data,
    )
